file.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	// check that the length of data directory entries is large.
+	// enough to include the COM descriptor directory.
+	if dd_length < pe.IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR+1 {
+		return nil, fmt.Errorf("data directory entries length (%d) is less than minimum length (%d) to include the COM descriptor directory", dd_length, pe.IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR)
+	}
+
+	clihdr, err := readCLIHeader(pefile, pe64)
+	if err != nil {
+		return nil, err
+	}
+	mdhdr, streams, err := readMetadata(pefile, clihdr.Metadata.VirtualAddress)
+	if err != nil {
+		return nil, err
+	}
+	return &File{
+		CLIHeader:      *clihdr,
+		MetadataHeader: *mdhdr,
+		Streams:        streams,
+	}, nil
+}
+
+// readCLIHeader reads the CLI header from pefile info f.
+func readCLIHeader(pefile *pe.File, pe64 bool) (*CLIHeader, error) {
+	// grab the com descriptor data directory entry.
+	var comdd pe.DataDirectory
+	if pe64 {
+		comdd = pefile.OptionalHeader.(*pe.OptionalHeader64).DataDirectory[pe.IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR]
+	} else {
+		comdd = pefile.OptionalHeader.(*pe.OptionalHeader32).DataDirectory[pe.IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR]
+	}
+
+	// figure out which section contains the COM descriptor directory table
+	var ds *pe.Section
+	for _, s := range pefile.Sections {
+		if s.VirtualAddress <= comdd.VirtualAddress && comdd.VirtualAddress < s.VirtualAddress+s.VirtualSize {
+			ds = s
+			break
+		}
+	}
+
+	// didn't find a section.
+	if ds == nil {
+		return nil, errors.New("COM descriptor directory table is missing")
+	}
+
+	// read COM descriptor directory table.
+	d, err := ds.Data()
+	if err != nil {
+		return nil, err
+	}
+
+	// seek to the virtual address specified in the COM descriptor data directory.
+	d = d[comdd.VirtualAddress-ds.VirtualAddress:]
+
+	var hdr CLIHeader
+	// verify there is enough space for the CLI header.
+	cliHeaderSize := binary.Size(hdr)
+	if len(d) < cliHeaderSize {
+		return nil, fmt.Errorf("COM descriptor data directory size (%d) is less than minimum size (%d) for CLI header", len(d), cliHeaderSize)
+	}
+
+	// parse CLI header.
+	hdr.Size = binary.LittleEndian.Uint32(d[0:4])
+	if hdr.Size < uint32(cliHeaderSize) {
+		return nil, fmt.Errorf("size of CLI header (%d) is not the one defined in ECMA335 6th edition, section 11.25.3.3 (%d)", hdr.Size, cliHeaderSize)
+	}
+	hdr.MajorRuntimeVersion = binary.LittleEndian.Uint16(d[4:6])
+	hdr.MinorRuntimeVersion = binary.LittleEndian.Uint16(d[6:8])
+	hdr.Metadata = readDataDirectory(d[8:16])
+	hdr.Flags = COMIMAGE_FLAGS(binary.LittleEndian.Uint32(d[16:20]))
+	hdr.EntryPointToken = binary.LittleEndian.Uint32(d[20:24])
+	hdr.Resources = readDataDirectory(d[24:32])
+	hdr.StrongNameSignature = readDataDirectory(d[32:40])
+	hdr.CodeManagerTable = readDataDirectory(d[40:48])
+	hdr.VTableFixups = readDataDirectory(d[48:56])
+	hdr.ExportAddressTableJumps = readDataDirectory(d[56:64])
+	hdr.ManagedNativeHeader = readDataDirectory(d[64:72])

```

Comment:
>Curious why the manual-looking numbers rather than using consume-style funcs?

This header length is fixed, it does not have any variadic field, and in these cases I prefer the manual decoding. But just the next function uses (`readMetadata`) uses a consume-style decoding, and it looks weird to combine both styles. I'll refactor the first one for the sake of consistency.